#166272336 Feature Sign out user

despeauxz opened the pull request:
#### What does this PR do?
This PR adds functionality for a user to sign out 
#### Description of Task to be completed?
- Created a DroppedToke table for dropped tokens
- Create an authorize middleware that checks the token if it has been dropped or is not provided
#### How should this be manually tested?
Navigate to the endpoint `http://localhost:3000/api/v1/auth/logout`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166272336](https://www.pivotaltracker.com/story/show/166272336)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

despeauxz committed:
ft-[166272336]: signout user

- drops user token

[Delivers #166272336]

fidelisojeah commented on server/middlewares/Authorization.js:
```diff
@@ -0,0 +1,72 @@
+    */
+  static getToken(req) {
+    const bearerToken = req.headers.authorization;
+    const token = bearerToken && bearerToken.replace('Bearer ', '');
```
token would be `false` if `bearerToken` does not exist

You should properly change this to something like:

```javascript
if (bearerToken && bearerToken.startsWith('Bearer ')){
  const token = bearerToken.replace('Bearer ', '');
  return token
} else if (bearerToken){
  return bearerToken; // This part is not entirely needed (only gives the option for the token to be set in the header directly
}else {
  throw new Error('Invalid token supplied');
}
```

then move the `const token = Authorization.getToken(req);` into the `try/catch` block

despeauxz committed:
ft-[166272336]: signout user

- drops user token

[Delivers #166272336]

fidelisojeah commented on server/middlewares/Authorization.js:
```diff
@@ -0,0 +1,77 @@
+    */
+  static async authorize(req, res, next) {
+    const token = Authorization.getToken(req);
+    if (!token) return res.status(400).json({ status: 400, error: 'No token specified' });
```
```javascript
errors: {
  global: 'Invalid token supplied: format Bearer <token>'
}
```

despeauxz committed:
feat-[166272336]: signout user

- drop user token
- logout user
- authorize user token
- test the logout endpoint

[Finishes #166272336]

despeauxz committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272336-ft-signout-user

despeauxz committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272336-ft-signout-user

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -150,6 +150,27 @@ class UserController {
+        status: 201, message: 'You are now logged out'
+      });
+    } catch (error) {
+      return res.status(401).json({
```
similar error format issue here

Change the error format here

fidelisojeah commented on server/helpers/Token.js:
```diff
@@ -23,27 +26,56 @@ class Token {
+        }
+      });
+      if (droppedToken) {
+        return res.status(401).json({
```
wrong error format

fidelisojeah commented on server/models/droppedtoken.js:
```diff
@@ -0,0 +1,12 @@
+    updatedAt: DataTypes.DATE
+  }, {});
+
+  // DroppedToken.associate = (models) => {
```
remove commented code

fidelisojeah reviewed (changes requested):
Please pull from develop and correct current issues

GerrardE commented:
Good job. Do you mind updating the API docs?

EMacco reviewed (changes requested):
Great work Malik, although I think should could also consider handling the error if a token is not provided at all. You could return a global error `You are not logged in`. Currently it returns `Cannot read property 'startsWith' of undefined`
<img width="973" alt="Screenshot 2019-06-11 at 10 47 05 PM" src="https://user-images.githubusercontent.com/20377428/59309970-e2d66900-8c9c-11e9-902e-fab8a088b38f.png">

Also, considering the fact that you now have Bearer appended to the token, you could update the login and register routes so that it returns the token with 'Bearer ' appended to it. 
<img width="973" alt="Screenshot 2019-06-11 at 10 49 40 PM" src="https://user-images.githubusercontent.com/20377428/59310047-1913e880-8c9d-11e9-85e5-f66be06fc3b6.png">

fidelisojeah commented:
> Great work Malik, although I think should could also consider handling the error if a token is not provided at all. You could return a global error `You are not logged in`. Currently it returns `Cannot read property 'startsWith' of undefined`
> <img alt="Screenshot 2019-06-11 at 10 47 05 PM" width="973" src="https://user-images.githubusercontent.com/20377428/59309970-e2d66900-8c9c-11e9-902e-fab8a088b38f.png">
> 
> Also, considering the fact that you now have Bearer appended to the token, you could update the login and register routes so that it returns the token with 'Bearer ' appended to it.
> <img alt="Screenshot 2019-06-11 at 10 49 40 PM" width="973" src="https://user-images.githubusercontent.com/20377428/59310047-1913e880-8c9d-11e9-85e5-f66be06fc3b6.png">


please do not append `Bearer ` to the token on generation

However, all other feedback is okay

despeauxz reviewed (commented):
@fidelisojeah The story does not require I document the endpoint with swagger and the only field required is the req.header.authorization. As for the other feedback by @EMacco, that doesn't apply cos I just tested again to be sure and no such error. But will definitely update the error response feedback

despeauxz committed:
Merge branch '166272336-ft-signout-user' of https://github.com/andela/kingsmen-ah-backend into 166272336-ft-signout-user
